savant/api/builder.py

```diff
@@ -0,0 +1,166 @@
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+from savant_rs.primitives import (
+    Attribute,
+    AttributeValue,
+    IdCollisionResolutionPolicy,
+    VideoFrame,
+    VideoFrameContent,
+    VideoObject,
+)
+from savant_rs.primitives.geometry import BBox, RBBox
+from savant_rs.video_object_query import IntExpression, MatchQuery
+
+from savant.api.constants import DEFAULT_NAMESPACE, DEFAULT_TIME_BASE
+from savant.meta.constants import UNTRACKED_OBJECT_ID
+
+
+def build_video_frame(
+    source_id: str,
+    framerate: str,
+    width: int,
+    height: int,
+    pts: int,
+    keyframe: bool,
+    content: Optional[Union[bytes, Tuple[str, Optional[str]]]],
+    codec: Optional[str] = None,
+    dts: Optional[int] = None,
+    duration: Optional[int] = None,
+    objects: Optional[List[Dict[str, Any]]] = None,
+    tags: Optional[Dict[str, Union[bool, int, float, str]]] = None,
+    time_base: Tuple[int, int] = DEFAULT_TIME_BASE,
+) -> VideoFrame:
+    if content is None:
+        content = VideoFrameContent.none()
+    elif isinstance(content, bytes):
+        content = VideoFrameContent.internal(content)
+    else:
+        content = VideoFrameContent.external(*content)
+    frame = VideoFrame(
+        source_id=source_id,
+        framerate=framerate,
+        width=width,
+        height=height,
+        codec=codec,
+        content=content,
+        keyframe=keyframe,
+        pts=pts,
+        dts=dts,
+        duration=duration,
+        time_base=time_base,
+    )
+    if objects:
+        add_objects_to_video_frame(frame, objects)
+    if tags:
+        add_tags_to_video_frame(frame, tags)
+
+    return frame
+
+
+def add_objects_to_video_frame(
+    frame: VideoFrame,
+    objects: List[Dict[str, Any]],
+):
+    parents = {}
+    for obj_id, obj in enumerate(objects):
+        video_object = build_video_object(obj_id, obj)
+        frame.add_object(video_object, IdCollisionResolutionPolicy.Error)
+        track_id = video_object.get_track_id()
+        if track_id is not None:
+            parents[
+                (video_object.namespace, video_object.label, track_id)
+            ] = video_object
+
+    for obj_id, obj in enumerate(objects):
+        parent_id = obj.get('parent_object_id')
+        parent_model_name = obj.get('parent_model_name')
+        parent_label = obj.get('parent_label')
+        if parent_id is None or parent_model_name is None or parent_label is None:
+            continue
+        frame.set_parent(
```

Comment:
@tomskikh I added `set_parent_by_id(child_id, parent_id)` which does it more efficiently. Raises an exception if no parent or child exists. Available in `0.1.38`.

Reply:
Updated savant-rs to 0.1.38